### ISSUE TITLE
False Positive: Cases with effectively stable or controlled lock objects are still treated as synchronization on non-final fields.

### ISSUE THREAD

**Issue by Carlson-JLQ** [google/error-prone#5644]
# False Positive: Cases with effectively stable or controlled lock objects are still treated as synchronization on non-final fields.

Version
2.48.0

## Checker
- Checker id: `SynchronizeOnNonFinalField`
- Checker description: This checker detects synchronization on a non-final field that is not safely exempted by being local, static, final, lazily initialized, or belonging to specific I/O classes.

## Description of the false positive
`diff/output` marks 3 false positives for `SynchronizeOnNonFinalField`. Across these samples, the checker tends to fire on the outer syntax while missing code-level cues that make the pattern harmless, exempted, or at least much weaker than a real bug. The recurring themes are synchronization patterns.

### `NegCase2.java`
Source: `output/sonarqube/SynchronizedFieldAssignmentCheck/src/main/java/scensct/core/neg/NegCase2.java`

This is a concurrency-pattern sample where the same bug or non-bug is expressed through synchronization primitives rather than a single canonical syntax form. A human reader would likely classify this as an allowed or intentionally structured pattern rather than a bug. That matches the benchmark's stated intent: Synchronizing on an unresolved field symbol should not be flagged as synchronization on a non-private final field assigned within the block.

```java
// Synchronizing on an unresolved field symbol should not be flagged as synchronization on a non-private final field assigned within the block.
package scensct.core.neg;

import external.UnknownType;

public class NegCase2 {
    private UnknownType unknownField;
    public void method() {
        synchronized (unknownField) { // Field symbol is unknown/unresolved.
            // No assignment inside.
        }
    }
}
```

### `NegCase4.java`
Source: `output/sonarqube/SynchronizedFieldAssignmentCheck/src/main/java/scensct/core/neg/NegCase4.java`

This is a concurrency-pattern sample where the same bug or non-bug is expressed through synchronization primitives rather than a single canonical syntax form. A human reader would likely classify this as an allowed or intentionally structured pattern rather than a bug. That matches the benchmark's stated intent: Synchronizing on a known field with no direct assignments to simple identifier variables inside the block should not be flagged.

```java
// Synchronizing on a known field with no direct assignments to simple identifier variables inside the block should not be flagged.
package scensct.core.neg;

public class NegCase4 {
    private Object lock = new Object(); // Known field owned by a type.
    
    public void method() {
        synchronized (lock) {
            // No assignment to any simple identifier variable (e.g., lock = ...).
            // Only method calls or other operations.
            System.out.println("No assignment");
        }
    }
}
```

### `NegCase5.java`
Source: `output/sonarqube/SynchronizedFieldAssignmentCheck/src/main/java/scensct/core/neg/NegCase5.java`

This is a concurrency-pattern sample where the same bug or non-bug is expressed through synchronization primitives rather than a single canonical syntax form. A human reader would likely classify this as an allowed or intentionally structured pattern rather than a bug. That matches the benchmark's stated intent: Synchronizing on a known field where assignments inside the block are to a different variable, not the lock field, should not be flagged.

```java
// Synchronizing on a known field where assignments inside the block are to a different variable, not the lock field, should not be flagged.
package scensct.core.neg;

public class NegCase5 {
    private Object lock = new Object(); // Known field owned by a type.
    private Object otherField = new Object();
    
    public void method() {
        synchronized (lock) {
            otherField = new Object(); // Assignment to a different field, not 'lock'.
        }
    }
}
```

## Cause analysis
The rule is documented as follows: This checker detects synchronization on a non-final field that is not safely exempted by being local, static, final, lazily initialized, or belonging to specific I/O classes.

In this set the problem shows up around synchronization patterns. The implementation appears to lean too heavily on surface indicators and not enough on what the code actually does.

That is why these reports read like over-approximation. The code often retains a suspicious signature or control-flow skeleton, but once you read the body, the benchmark's claimed exemption is visible in the source itself.

For `SynchronizeOnNonFinalField`, the practical improvement would be to model those source-level exemptions more explicitly instead of reporting every syntactic near-match.

## References
None known.
